Admin/components/AdminSite/Profile.php

```diff
@@ -164,6 +167,40 @@ protected function buildInternal()
 			$confirm_password->hasMessage()) {
 			$this->ui->getWidget('change_password')->open = true;
 		}
+
+		$this->buildGoogle2fa();
+	}
+
+	// }}}
+	// {{{ protected function buildGoogle2fa()
+
+	protected function buildGoogle2fa()
+	{
+		if ($this->app->isGoogle2faEnabled()) {
+			$google2fa = new Google2FA();
+			if ($this->data_object->google_2fa_secret === null) {
+				$google2fa_secret = $google2fa->generateSecretKey();
+				$this->data_object->google_2fa_secret = $google2fa_secret;
+				$this->data_object->save();
+			}
+
+			if ($this->data_object->google_2fa_enabled) {
+				$this->ui->getWidget('google_2fa_enabled_note')->visible = true;
+			} else {
+				$qr_code_url = $google2fa->getQRCodeUrl(
+					$this->app->config->admin->google_2fa_domain,
+					$this->data_object->email,
+					$this->data_object->google_2fa_secret
+				);
+
+				$img_tag = new SwatHtmlTag('img');
+				$img_tag->src = 'https://chart.googleapis.com/chart'.
```

Comment:
this is unfortunately deprecated. I think we'll need to replace with Bacon https://github.com/antonioribeiro/google2fa#baconqrcode